slsa/provenance_test.go

```diff
@@ -0,0 +1,59 @@
+package slsa
+
+import (
+	"reflect"
+	"testing"
+
+	intoto "github.com/in-toto/in-toto-golang/in_toto"
+	slsa "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v0.2"
+
+	"github.com/slsa-framework/slsa-github-generator/github"
+)
+
+func TestHostedActionsProvenance(t *testing.T) {
+	testCases := []struct {
+		name     string
+		r        WorkflowRun
+		expected *intoto.ProvenanceStatement
+		err      bool
```

Comment:
how about making this an `error` type? Then you can test the error is of the expected type, see https://github.com/slsa-framework/slsa-verifier/blob/main/main_test.go#L306
This has been useful for me in the past, I caught some non-expected errors :-)

Reply:
I can do that though I don't think I'm using any particular types at the moment (just whatever type happens to be returned by fmt.Errorf). I suppose I could return different types but didn't necessarily think it was worth it as it really wouldn't be used outside of tests. Maybe I'll just return some unexported types for now.

Reply:
I think it's still useful to test that the internal function work as expected. I tend do do this:
```
var errSomeName = errors.New("some msg)

fun() {
  return fmt.Errorf("%w", errSomeName)
}
```

Would this work?

Reply:
Yeah, I'll make this a bit better. At least for this file it's mute since it's not really used. I'll add some more tests over on RequestOIDCToken since that's really where the errors are produced.

In general, it's a bit annoying in Go to add custom error messages with your own types.

One can do:
```golang
if err := someFunc(); err != nil { 
  return fmt.Errorf("%w: %v", errSomeName, err)
}
```
and get the property that `errors.Is(err, errSomeName) == true` but then you loose the type of the error returned from someFunc and I'd not really consider that a good practice.

The "right" way to do it is to go full bore with your own wrapping error type but it gets cumbersome quick as you need to do that for every type you want to test for. We could have our own internal base wrapping type that can compare types and allow custom messages but it's maybe a bit of a pain to use: https://go.dev/play/p/6f1bDFwr8vj

Reply:
I think is to define our own types and use `errors.As` to compare is probably what we want so I'll take a stab at it.

Reply:
Something like this is what I came up with, but I'll hold off on adding this and just use static errors for now.
https://go.dev/play/p/H6mrl41DzAX

Reply:
Does  the following work?

```go
fmt.Errorf("%w: %w", errSomeName, err)
```

Reply:
Errorf only supports one wrapped error AFAICT
https://cs.opensource.google/go/go/+/refs/tags/go1.18:src/fmt/errors.go;l=17

But incidentally we could do something like wrap twice maybe.
```golang
fmt.Errorf("%w: %v", fmt.Errorf("%w", errSomeName), err)
```

The real trick comes when checking with either `errors.Is` which checks strict equality or `errors.As` which checks type equality. In any case, we can keep this in mind for future PRs. I think we should be good for this one.